src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -471,3 +471,84 @@ lemma ideal.homogeneous_hull_eq_Inf (I : ideal A) :
+variables [algebra R A] [decidable_eq ι]
+variables [canonically_ordered_add_monoid ι] [covariant_class ι ι has_add.add has_lt.lt]
+variables (𝒜 : ι → submodule R A) [graded_algebra 𝒜]
+
+/--
+For a graded ring `⨁ᵢ 𝒜ᵢ` graded by a `canonically_ordered_add_monoid ι`, the irrelevant ideal
+refers to `⨁_{i≥0} 𝒜ᵢ`, or equivalently `{a | the bot-th projection of a = 0}`.
+
+This definitoin is used in Proj construction where `ι` is always `ℕ` so the irrelevant ideal is
+simply elements with `0` as 0-th coordinate. But the notion of irrelevant ideal makes sense in a
+more general setting by defining it as the ideal of elements with `0` as i-th coordinate for all
+`i ≤ 0`.
+-/
+def homogeneous_ideal.irrelevant : homogeneous_ideal 𝒜 :=
+let f : A →+* A :=
+{ to_fun := λ a, proj 𝒜 0 a,
+  map_one' := by { rw [proj_apply, decompose_of_mem_same], exact set_like.graded_monoid.one_mem },
+  map_mul' := λ x y, begin
+    haveI : Π (i : ι) (x : (𝒜 i)), decidable (x ≠ 0) := λ _ _, classical.dec_pred _ _,
+    apply @submodule.supr_induction _ _ _ _ _ _ 𝒜
+        (λ x, proj 𝒜 0 (x * y) = proj 𝒜 0 x * proj 𝒜 0 y),
+    { rw (is_internal 𝒜).supr_eq_top, trivial, },
+    { intros i c hc,
+      apply @submodule.supr_induction _ _ _ _ _ _ 𝒜
+        (λ y, proj 𝒜 0 (c * y) = proj 𝒜 0 c * proj 𝒜 0 y),
+      { rw (is_internal 𝒜).supr_eq_top, trivial, },
+      { intros j c' hc',
+        have mem1 : c * c' ∈ 𝒜 (i + j) := set_like.graded_monoid.mul_mem hc hc',
+        by_cases ineq1 : i + j = 0,
+        { rw ineq1 at mem1,
+          rw [proj_apply, decompose_of_mem_same 𝒜 mem1],
+          rw ← bot_eq_zero at ineq1 ⊢,
+          have j_eq : j = ⊥,
+          { have ineq2 : ⊥ ≤ j := bot_le,
+            rw le_iff_lt_or_eq at ineq2,
+            cases ineq2,
+            { have ineq3 := add_lt_add_of_le_of_lt (bot_le : ⊥ ≤ i) ineq2,
+              rw ineq1 at ineq3,
+              exfalso,
+              exact lt_irrefl (⊥ + ⊥) (lt_of_lt_of_le ineq3 (bot_le : ⊥ ≤ ⊥ + ⊥)), },
+            { exact ineq2.symm }, },
+          rw bot_eq_zero at *,
+          rw j_eq at *,
+          rw [add_zero] at ineq1,
+          rw ineq1 at hc,
+          rw [proj_apply, proj_apply, decompose_of_mem_same 𝒜 hc, decompose_of_mem_same 𝒜 hc'] },
+        { rw [proj_apply, decompose_of_mem_ne 𝒜 mem1 ineq1],
+          have ineq2 : i ≠ 0 ∨ j ≠ 0,
+          { by_contra' rid,
+            rw [rid.1, rid.2, add_zero] at ineq1,
+            exact ineq1 rfl, },
+          cases ineq2,
+          { simp only [proj_apply, decompose_of_mem_ne 𝒜 hc ineq2, zero_mul] },
+          { simp only [proj_apply, decompose_of_mem_ne 𝒜 hc' ineq2, mul_zero] } } },
+      { rw [mul_zero, map_zero, mul_zero], },
+      { rintros d e (hd : _ = _ * _) (he : _ = _ * _),
+        rw [mul_add, map_add, hd, he, map_add, mul_add] }, },
+    { rw [zero_mul, map_zero, zero_mul] },
+    { rintros a b (ha : _ = _ * _) (hb : _ = _ * _),
+      rw [add_mul, map_add, ha, hb, map_add, add_mul], }
+  end,
```

Comment:
I think you should extract this to a definition `proj_zero_ring_hom`, and it should maybe have type `A →+* 𝒜 0`

Reply:
I changed to a version with type `A →+* 𝒜 0`, but it made
```lean
lemma homogeneous_ideal.mem_irrelevant_iff (a : A) :
  a ∈ homogeneous_ideal.irrelevant 𝒜 ↔ proj 𝒜 0 a = 0 :=
```
no longer definitionally true.
Is there any particular reason why `A →+* 𝒜 0` is better than `A →+* A`